beacon_node/beacon_chain/src/canonical_head.rs

```diff
@@ -654,11 +654,11 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
                     })
                 })
                 .and_then(|mut snapshot| {
-                    // Regardless of where we got the state from, attempt to build the committee
-                    // caches.
+                    // Regardless of where we got the state from, attempt to build all the
+                    // caches except the tree hash cache.
                     snapshot
                         .beacon_state
-                        .build_all_committee_caches(&self.spec)
```

Comment:
It may be a little heavy handed to build the exit cache here as well, but I don't feel too strongly about that. IIRC we don't *always* need the exit cache to process a block.

Reply:
I figure it's quasi-equivalent to building it on-demand when it is eventually required. If we process a block and it becomes head, chances are its descendants will remain canonical and the exit cache will be required eventually, at which point we've paid once to build it (when the first ancestor was processed). The only case where I think it'll hurt is if there are lots of re-orgs that miss the snapshot cache and load a fresh state without any caches. We hope that these are rare. These re-orgs and these sorts of caching issues are also mitigated by `tree-states`, where all the states share common CoW caches rooted at the finalized state.

Reply:
My (potentially shoddy) understanding is that the exit cache is only required when we're processing exits, which is relatively infrequently. However, given that:

- The cache we're building is on the head state.
- The head state usually maintains it's caches (i.e., they aren't dropped and rebuilt often)
- Trying to build an already-built exit cache is cheap.

I'm happy with this. I'll slap another approval!